### PR TITLE
Add mypy static type checker

### PR DESCRIPTION
**Pull request by samuelrince**
Many changes in this PR to make sure we properly type the package.

Each commit represents a fix of one "type" of issues related to mypy error reports.

Implements #90 

Also based on https://github.com/NP4567-dev/ecologits/tree/feature/mypy

### CODE REVIEW

File: ecologits/model_repository.py [hunk 1]
```diff
@@ -92,6 +93,8 @@ def from_json(cls, filepath: Optional[str] = None) -> "ModelRepository":
         with open(filepath) as fd:
             data = json.load(fd)
             mf = Models.model_validate(data)
+        if mf.models is None:
+            raise ValueError("Cannot initialize and empty model repository.")
```
Comment (NP4567-dev) [inline ecologits/model_repository.py:97]: nitpick: "Cannot initialize and empty model repository." --> "Cannot initialize on an empty model repository." ?

File: ecologits/impacts/modeling.py [hunk 2]
```diff
@@ -22,8 +25,8 @@ class Impact(BaseModel):
     value: ValueOrRange
     unit: str
 
-    def __add__(self, other: "Impact") -> "Impact":
-        if not isinstance(other, Impact):
+    def __add__(self: Impact, other: "BaseImpact") -> Impact:
+        if not isinstance(other, BaseImpact):
```
Comment (NP4567-dev) [inline ecologits/impacts/modeling.py:29]: question: as the type is checked and an error raised, we could just have a more generic "object" type?
I think I had a "not compliant with Liskov principles" on this one from mypy

**Review by NP4567-dev — APPROVED** [review]
👍

File: ecologits/impacts/modeling.py (hunk 2 above)
Comment (samuelrince) [reply ecologits/impacts/modeling.py:29]: Yes I agree it makes more sense if we compare to comparison methods.

**Pull request merged**